### ISSUE TITLE
finalize() without drop()?

### ISSUE THREAD

**Issue by arjennienhuis** [parallaxsecond/rust-cryptoki#150]
The only way I can find to "gracefully" stop another thread that is blocking on `C_WaitForSlotEvent` is to call `C_Finalize`.

`Pkcs11::finalize()` is a no-op and `Pkcs11Impl::finalize()` is only called on `drop()`. I cannot drop the library if it is being used in another thread.

Is it ok to let `Pkcs11::finalize()` call `Pkcs11Impl::finalize()`?

We also need to fix `Pkcs11::is_initialized()` but that is a separate issue.

**Comment by wiktor-k** [comment]
Hmm... Yeah, it looks problematic. I'm not the author of this code but I assume the intention is to allow multiple threads access through Arc to the Impl struct and finalize when all threads are done.

And your point is that you want to call finalize earlier? The problematic part is, the pkcs11 struct assumes that what it holds is an initialized object before finalization. Would it be possible for your code to get all threads to drop their references? That'd deterministically call finalize when the last reference is dropped.

**Comment by hug-dev** [comment]
> The problematic part is, the pkcs11 struct assumes that what it holds is an initialized object before finalization.

Agreed, if we were to change `Pkcs11::finalize` to only take a reference like the other functions and call `Pkcs11Impl::finalize` it would mean that users would still be able to use the other PKCS11 function and get the NotInitialized error everytime. I guess this is not a big safety problem but reducing those kind of errors is maybe still very nice to have?

Tell us if you are still having issues for this and if there are no other ways we can maybe think of something!
